### DISCUSSION TITLE
restrictTranslate only for left and right side of paper

### DISCUSSION THREAD

**Discussion by TobiasRama** [clientIO/joint#1923] [Q&A] [answered]
The option restrictTranslate of a paper restricts the translation of elements for all 4 sides. Is it possible to do so only for the left and right side?

**Comment by kumilingus** [comment, marked as answer]
The `restrictTranslate` option can be also defined as a rectangle or via a function (see [docs](https://resources.jointjs.com/docs/jointjs/v3.6/joint.html#dia.Paper.prototype.options.restrictTranslate)).

As a rectangle:
```js
const left = 0;
const right = 500;
paper.options.restrictTranslate = {
  x: left,
  y: -Number.MAX_SAFE_INTEGER,
  width: right - left,
  height: 2 * Number.MAX_SAFE_INTEGER
}
```

As a function:
```js
const left = 0;
const right = 500;
paper.options.restrictTranslate = () => ((x,y) => ({ x: Math.max(Math.min(x, right), left), y }))
```

**Reply by TobiasRama** [reply]
Thanks for helping me out, but this caused another problem in my application. I use shapes with a label on the outside like in Discussion https://github.com/clientIO/joint/discussions/1915. The problem is: the translation stops because of the model, but I am still allowed to drag the label outside of the paper. Is there a way to stop the translation based on the boundary box of the shape?

**Reply by kumilingus** [reply]
If the label is below the element you can do this:
```js
restrictTranslate: (elementView) => {
    const viewBBox = elementView.getNodeBBox(elementView.el);
    const modelBBox = elementView.model.getBBox();
    const area = paper.getArea();
    area.height -= viewBBox.height - modelBBox.height;
    return area;
}
```
